BRANCHING.md

```diff
@@ -0,0 +1,214 @@
+```
+
+Even though a commit in `release/v3.1` may follow two paths into master branch,
+it does not conflict with itself under a `git merge`. (This is another reason
+why a normal git merge should always be preferred when merging two long-lived branches.)
+
+Comparisons
+-----------
+
+Historically, `mobilecoin.git` did not use merging at all when handling release branches.
+
+* We branch for release e.g. `release/v2.0` of off `master`.
+* Developers (usually) make bug fix commits in release branch, or,
+  cherry-pick them back to release branch.
+* We tag release candidates and releases on the release branch.
+* If a point release is needed off of a major release, then we branch e.g.
+  `release/v2.1` off of `release/v2.0` and make more commits.
+* After the release is done, the release branch is abandoned and nothing more is done with it.
+
+In this model:
+
+* Release branches are never merged anywhere
+* Any changes that need to be in master and release (which is most bug fixes)
+  get cherry-picked.
+
+This is an example of what this looks like:
+
+[![](https://mermaid.ink/img/pako:eNqdVMtugzAQ_BW05_QRCHXh3KqnnnrlsrEXsIIBOSZqFOXf65C0CtSmoT5hZmY9s34cgDeCIIXAjkKaN41tmdVBP3ijlDTu2VpjzctAU0W4pYddOKQFUqQZ4CpOMnAh6yf2_AsxWFhoF94vr6CS-KbpTKBQ1h5jJWm9v2sl3_iX9ZJdTrzSs5Xp0Hz56AktGIv-Sja2N9LMNPTT0XDuuq4U_zkXkasRFMeRu0U5Y6t5m38l_kySgXaYyFPa4XSsHNa95YRMtnaUfsLIYBOjudfiMoMFKNKWJ-w1P5zQDExJijJI7aegHLvKnEofLRU703zsaw6p0R0toGsFGnqRWGhUkOZYbe1fEtI0-v38dPQvyDfztUcuxOMXurlSCQ?type=png)](https://mermaid.live/edit#pako:eNqdVMtugzAQ_BW05_QRCHXh3KqnnnrlsrEXsIIBOSZqFOXf65C0CtSmoT5hZmY9s34cgDeCIIXAjkKaN41tmdVBP3ijlDTu2VpjzctAU0W4pYddOKQFUqQZ4CpOMnAh6yf2_AsxWFhoF94vr6CS-KbpTKBQ1h5jJWm9v2sl3_iX9ZJdTrzSs5Xp0Hz56AktGIv-Sja2N9LMNPTT0XDuuq4U_zkXkasRFMeRu0U5Y6t5m38l_kySgXaYyFPa4XSsHNa95YRMtnaUfsLIYBOjudfiMoMFKNKWJ-w1P5zQDExJijJI7aegHLvKnEofLRU703zsaw6p0R0toGsFGnqRWGhUkOZYbe1fEtI0-v38dPQvyDfztUcuxOMXurlSCQ)
+
+The main drawbacks of this are:
+
+* Changes are being cherry-picked in two directions, which can be confusing and
+  can lead to additional git conflicts.
+* Changes are sometimes cherry-picked from release to master in reverse order
+  to how they occurred on release. This can create confusion and additional git conflicts.
+  Sometimes, git ends up pulling "fragments" of the earlier commit that was not cherry-picked
+  yet forwards into master, which is very confusing for developers and reviewers because
+  seemingly unrelated changes appear in the diff.
+* Sometimes changes in release are forgotten about and never cherry-picked to master.
+  This leads to regressions in the next release. Because we never merge the release
+  branch back to master, and all the cherry-picks have different hashes from their
+  originals, it's very hard to ever be sure that we got all the changes.
+
+This merge-based policy hopes to rectify these issues.
+
+Why squash merge at all
+-----------------------
+
+The best argument for why topic branches should be squash merged is that,
```

Comment:
It also has the major benefits of ensuring the history of any base branch (`master`/`main`/`release/*`) is readable, and discourages their histories from being filled with unhelpful commit messages like "debugging" and "fixes". It also dramatically improves the usefulness of tools that rely on git-blame (e.g. GitLense in VSCode), as the commits they point to offer an insight into the the reason for the change from their title.

Reply:
Having worked in projects for years that don't do any squash merging, I guess I don't think this is a benefit of squash merging.

In projects that don't do any squash merging, you don't allow people to merge a PR with commit messages like "debugging" and "fixes". They have to either use `git commit --amend` so the fixups are part of the previous commit, or `git rebase` on their branch and squash and edit things to make their commit history useful before merging.

Anyways, you can still navigate to the merge commit and read the same commit message you would have gotten in a squash merge. It's only like another click or two.